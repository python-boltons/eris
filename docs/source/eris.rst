eris package
============

.. automodule:: eris
   :members:
   :undoc-members:
   :show-inheritance:
