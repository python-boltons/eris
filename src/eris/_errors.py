"""Custom Exceptions live here."""

from __future__ import annotations

import traceback
from typing import (
    Final,
    Iterator,
    List,
    Literal,
    Optional,
    Tuple,
    TypedDict,
    Union,
)

from ion import efill
from metaman import Inspector, cname
from typist import E, T


Null = Literal["null"]
Nullable = Union[Null, T]
ErisErrorChain = List["ErisErrorDict"]
ExcInfoTuple = Tuple[str, str, Nullable[str]]

FIRST_EXC_IS_WRONG_TYPE: Final = (
    "Logic error. The first exception returned by iterating over this"
    " exception should be THIS exception."
)
NULL: Final[Null] = "null"


class ExcInfo(TypedDict):
    """Represents a single exception."""

    exc_type: str
    exc_value: str
    exc_msg: str


class ErisErrorDict(TypedDict):
    """An Error type represented as a dictionary."""

    # exception info
    exc_info: ExcInfo

    # metadata
    lineno: int
    module_name: str
    func_name: str
    file_name: str

    # optional traceback stack + exception that caused it
    stack: List[str]
    caused_by: Nullable[List[ExcInfoTuple]]


class ErisError(Exception):
    """Custom general-purpose exception."""

    def __init__(self, emsg: str, up: int = 0) -> None:
        self.inspector = Inspector(up=up + 1)
        super().__init__(emsg)

    def __str__(self) -> str:  # noqa: D105
        return self.__repr__()

    def __repr__(self) -> str:  # noqa: D105
        return self._repr()

    def _repr(self, width: int = 80) -> str:
        """
        Format error to width.  If width is None, return string suitable for
        traceback.
        """
        super_str = super().__str__()

        emsg = efill(super_str, width, indent=2)
        return "{}::{}::{}::{}{{\n{}\n}}".format(
            cname(self),
            self.inspector.module_name,
            self.inspector.function_name,
            self.inspector.line_number,
            emsg,
        )

    def __iter__(self) -> Iterator["BaseException"]:  # noqa: D105
        yield self

        e = self.__cause__
        while e:
            yield e
            e = e.__cause__

    def chain(self, other: Exception) -> "ErisError":
        """Chains this exception to another."""
        return _chain_errors(self, other)

    def to_json(self) -> ErisErrorChain:
        """Converts this error into a list of dictionaries.

        This list is JSON serializable and is composed of data on this
        exception and any that are chained to it.

        NOTE: The list is sorted from last-to-first exception to be raised
        (i.e. this exception, the exception that caused this exception,
        etc...).
        """
        result = []
        last_stack: Optional[List[str]] = None
        last_caused_by: Optional[List[ExcInfoTuple]] = None
        for error in self:
            if isinstance(error, ErisError):
                caused_by = last_caused_by = []
                stack = last_stack = list(error.inspector.lines)

                exc_info: ExcInfo = dict(
                    exc_type=repr(type(error)),
                    exc_value=repr(error),
                    exc_msg=error.args[0],
                )
                eris_error_dict: ErisErrorDict = dict(
                    exc_info=exc_info,
                    lineno=error.inspector.line_number,
                    module_name=error.inspector.module_name,
                    func_name=error.inspector.function_name,
                    file_name=error.inspector.file_name,
                    stack=stack,
                    caused_by=caused_by,
                )
                result.append(eris_error_dict)
            else:
                assert last_stack is not None, FIRST_EXC_IS_WRONG_TYPE
                assert last_caused_by is not None, FIRST_EXC_IS_WRONG_TYPE

                # extend the last Error's 'caused_by' list with this
                # Exception's info...
                exc_info_tuple: ExcInfoTuple = (
                    repr(type(error)),
                    repr(error),
                    str(error.args[0]) if error.args else NULL,
                )
                last_caused_by.append(exc_info_tuple)

                # extend the stack by using lines from this Exception's
                # traceback...
                if tb := error.__traceback__:
                    last_stack.extend(traceback.extract_tb(tb).format())

        return result


def _chain_errors(e1: E, e2: Optional[Exception]) -> E:
    """Chain two exceptions together.

    This is the functional equivalent to ``raise e1 from e2``.

    Args:
        e1: An exception.
        e2: The exception we want to chain to ``e2``.

    Returns:
        ``e1`` after chaining ``e2`` to it.
    """
    e: BaseException = e1
    cause = e.__cause__
    while cause:
        e = cause
        cause = e.__cause__
    e.__cause__ = e2
    return e1
