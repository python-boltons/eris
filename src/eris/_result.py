"""Result Return Type / Error-Handling Implementation

This module implements an error-handling model that is heavily influenced by
the Rust programming language's Result type [1] and an article titled
"Python: better typed than you think" [2]. This model is used as the default
error-handling model in some modern programming languages (e.g. Go and Rust),
but has also been used successfully for years by functional programming
languages (e.g. Haskell).

The model revolves around the use of the Result type, which, in turn, is
always either an Ok instance or an Err instance.

[1]: https://doc.rust-lang.org/std/result
[2]: https://beepb00p.xyz/mypy-error-handling.html#coderef-throw_exc

Examples:
    Consider a function which has a known error-state. If an error does occur,
    we want to propagate the error to the caller somehow. We can rely on
    Python's built-in exceptions to handle this propagation, like so:

        def do_stuff() -> str:
            # do some stuff which sets the 'status' variable
            if status == SUCCESS:
                return "SUCCESS!"
            else:
                raise SomeError("<Error Context>")

    This method couples the caller to the `do_stuff()` function's
    implementation, however, since the caller MUST know that `do_stuff()` might
    raise a SomeError exeption before calling `do_stuff()`. Furthermore, if the
    caller does NOT handle this exception, there is no warning; SomeError will
    continue to propagate up the call-stack until it crashes the program.

    This module attempts to offer a safer approach. Using the `Result` return
    type, we might define `do_stuff()` like this:

        def do_stuff() -> Result[str, SomeError]:
            # do some stuff which sets the 'status' variable
            if status == SUCCESS:
                return Ok("SUCCESS!")
            else:
                e = SomeError("<Error Context>")
                return Err(e)

    This approach has the benefit of being type-safe: A function that calls
    `do_stuff()` MUST check for errors. This is enforced by type-checking tools
    like mypy, but is also made fairly obvious to the caller given the return
    type of `do_stuff()`. To demonstrate, let us now consider how a client
    might go about calling `do_stuff()`:

        def main() -> int:
            msg_result = do_stuff()
            if isinstance(msg_result, Err):
                e = msg_result.err()
                logger.error("An error occurred while doing stuff: %r", e)
                return 1

            msg = msg_result.ok()
            logger.info(msg)
            return 0

    Some might say a downside of this approach is that it requires you to write
    a lot of boilerplate error-handling logic. I would argue that this is yet
    another benefit of this approach, since dangerous code _should_ look
    dangerous. With that said, if we just want to crash the program on error,
    we could shorten the above `main()` function like so:

        def main() -> int:
            msg = do_stuff().unwrap()  # raises SomeError if an error occurs
            logger.info(msg)
            return 0
"""

from abc import ABC, abstractmethod
from dataclasses import InitVar, dataclass, field
from functools import wraps
from typing import (
    Any,
    Callable,
    Dict,
    Generic,
    NoReturn,
    Optional,
    Tuple,
    TypeVar,
    Union,
)

from ._errors import Error, ErrorChain


T = TypeVar("T")


class _ResultMixin(ABC, Generic[T]):
    def __bool__(self) -> NoReturn:
        raise ValueError(
            f"{self.__class__.__name__} object cannot be evaluated as a"
            " boolean. This is probably a bug in your code. Make sure you are"
            " either explicitly checking for Err results or using one of the"
            f" `Result.unwrap*()` methods:  {self!r}"
        )

    @abstractmethod
    def err(self) -> Optional[Error]:
        """Returns None if successful or an Exception type otherwise."""

    @abstractmethod
    def unwrap(self) -> T:
        """Returns real return type or raises an exception if unsuccessful."""


@dataclass(frozen=True)
class Ok(_ResultMixin[T]):
    """Ok result type.

    A value that indicates success and which stores arbitrary data for the
    return value.
    """

    _value: T

    @staticmethod
    def err() -> None:  # noqa: D102
        return None

    def ok(self) -> T:  # noqa: D102
        return self._value

    def unwrap(self) -> T:  # noqa: D102
        return self.ok()


@dataclass
class Err:
    """Err result type.

    A value that signifies failure and which stores arbitrary data for the
    error.
    """

    _error_or_string: InitVar[Union[str, Error]]
    _error: Error = field(init=False)

    def __post_init__(self, error_or_string: Union[str, Error]) -> None:
        """Set the `_error` attribute."""
        if isinstance(error_or_string, str):
            self._error = Error(error_or_string, up=2)
        else:
            self._error = error_or_string

    def err(self) -> Error:  # noqa: D102
        return self._error

    def unwrap(self) -> NoReturn:  # noqa: D102
        raise self.err()

    def chain(self, exc_or_err: Union[Exception, "Err"]) -> "Err":
        """Wraps another Exception object with this Exception object."""
        if isinstance(exc_or_err, Exception):
            other_exception = exc_or_err
        else:
            other_exception = exc_or_err.err()

        self.err().chain(other_exception)
        return self

    def to_json(self) -> ErrorChain:
        """A thin wrapper around self.err().to_json()."""
        return self.err().to_json()


# The 'Result' return type is used to implement an error-handling model heavily
# influenced by that used by the Rust programming language
# (see https://doc.rust-lang.org/book/ch09-00-error-handling.html).
Result = Union[Ok[T], Err]


def return_lazy_result(
    func: Callable[..., Result[T]]
) -> Callable[..., "LazyResult[T]"]:
    """Converts the return type of a function from result to a "lazy" result.

    In order to fetch the real return type from lazy_result, you must call
    lazy_result.result() or any other valid Result method [e.g.
    lazy_result.unwrap()].

    This decorator is useful when dealing with functions that return
    Result[None] (i.e. functions that are used soley for their
    side-effects), since it makes it harder to ignore potential errors.
    """

    @wraps(func)
    def wrapper(*args: Any, **kwargs: Any) -> LazyResult[T]:
        return LazyResult(func, *args, **kwargs)

    return wrapper


class LazyResult(_ResultMixin[T]):
    """See `help(return_lazy_result)`."""

    def __init__(
        self, func: Callable[..., Result[T]], *args: Any, **kwargs: Any
    ) -> None:
        self._func = func
        self._args: Tuple[Any, ...] = args
        self._kwargs: Dict[str, Any] = kwargs

        self._result: Optional[Result[T]] = None

    def result(self) -> Result[T]:
        """Retrieve the Result object corresponding with this LazyResult.

        Calls the function corresponding with this LazyResult (this function
        will only be called once, even if this method is called multiple times)
        and returns the same Result returned by that function.
        """
        if self._result is None:
            self._result = self._func(*self._args, **self._kwargs)
        return self._result

    def err(self) -> Optional[Error]:  # noqa: D102
        return self.result().err()

    def unwrap(self) -> T:  # noqa: D102
        return self.result().unwrap()
