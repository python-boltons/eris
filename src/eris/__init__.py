"""eris package

The Greek Goddess of Strife and Discord... The better way to handle, trace, and
log errors in Python.
"""

import logging as _logging

from ._errors import ErisError, ErisErrorChain, ErisErrorDict
from ._result import Err, LazyResult, Ok, Result, return_lazy_result


__all__ = [
    "ErisError",
    "ErisErrorChain",
    "ErisErrorDict",
    "Err",
    "LazyResult",
    "Ok",
    "Result",
    "return_lazy_result",
]

__author__ = "Bryan M Bugyi"
__email__ = "bryanbugyi34@gmail.com"
__version__ = "0.1.2"

_logging.getLogger(__name__).addHandler(_logging.NullHandler())
