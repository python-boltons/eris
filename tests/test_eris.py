"""Tests for the eris package."""

from __future__ import annotations

from typing import Final

from syrupy.assertion import SnapshotAssertion as Snapshot

from eris import ErisError, Err


ERROR_MSG: Final = "Something went wrong..."


def test_init_err_with_error() -> None:
    """Test the Err.err() method."""
    error = ErisError("test error")
    assert Err(error).err() == error


def test_init_err_with_string() -> None:
    """Test that instantiating an Err(foo) object, where 'foo' is a string."""
    err = Err(ERROR_MSG)
    error = err.err()
    assert error.args[0] == ERROR_MSG


def test_is_json__NO_CAUSE(snapshot: Snapshot) -> None:
    """Test the Err/Error.to_json() methods.

    NO ErisError chain and NO 'caused by' exceptions.
    """
    err = Err(ERROR_MSG)
    error = err.err()
    assert snapshot == error.to_json()


def test_is_json__ONE_CAUSE(snapshot: Snapshot) -> None:
    """Test the Err/Error.to_json() methods.

    NO ErisError chain and ONE 'caused by' exception.
    """
    try:
        x = 1 / 0
        print(x)
    except ZeroDivisionError as zero_div_error:
        err = Err(ERROR_MSG).chain(zero_div_error)
        assert snapshot == err.to_json()


def test_is_json__TWO_CAUSE(snapshot: Snapshot) -> None:
    """Test the Err/Error.to_json() methods.

    NO ErisError chain and TWO 'caused by' exceptions.
    """
    try:
        x = 1 / 0
        print(x)
    except ZeroDivisionError as zero_div_error:
        try:
            raise RuntimeError(
                "Why would we divide by zero?"
            ) from zero_div_error
        except RuntimeError as rt_error:
            err = Err(ERROR_MSG).chain(rt_error)
            assert snapshot == err.to_json()


def test_is_json__TWO_CAUSE_AND_RAISE_SELF(snapshot: Snapshot) -> None:
    """Test the Err/Error.to_json() methods.

    NO ErisError chain and TWO 'caused by' exceptions. This test case calls
    Error.to_json() after catching an Error exception.
    """
    try:
        x = 1 / 0
        print(x)
    except ZeroDivisionError as zero_div_error:
        try:
            raise RuntimeError(
                "Why would we divide by zero?"
            ) from zero_div_error
        except RuntimeError as rt_error:
            error = ErisError(ERROR_MSG)
            try:
                raise error from rt_error
            except ErisError as eris_error:
                assert snapshot == eris_error.to_json()


def test_is_json__CHAIN(snapshot: Snapshot) -> None:
    """Test the Err/Error.to_json() methods.

    Chain multiple Error objects together.
    """
    err1 = Err("Some FOO error has occurred.")
    err2 = Err("Some BAR error has occurred.").chain(err1)
    assert snapshot == err2.to_json()
