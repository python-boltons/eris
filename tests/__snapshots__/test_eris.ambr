# name: test_is_json__CHAIN
  <class 'list'> [
    <class 'dict'> {
      'caused_by': <class 'list'> [
      ],
      'exc_info': <class 'dict'> {
        'exc_msg': 'Some BAR error has occurred.',
        'exc_type': "<class 'eris._errors.ErisError'>",
        'exc_value': '
          ErisError::tests.test_eris::test_is_json__CHAIN::97{
            Some BAR error has occurred.
          }
        ',
      },
      'file_name': '/eris/tests/test_eris.py',
      'func_name': 'test_is_json__CHAIN',
      'lineno': 97,
      'module_name': 'tests.test_eris',
      'stack': <class 'list'> [
        '
              err2 = Err("Some BAR error has occurred.").chain(err1)
  
        ',
      ],
    },
    <class 'dict'> {
      'caused_by': <class 'list'> [
      ],
      'exc_info': <class 'dict'> {
        'exc_msg': 'Some FOO error has occurred.',
        'exc_type': "<class 'eris._errors.ErisError'>",
        'exc_value': '
          ErisError::tests.test_eris::test_is_json__CHAIN::96{
            Some FOO error has occurred.
          }
        ',
      },
      'file_name': '/eris/tests/test_eris.py',
      'func_name': 'test_is_json__CHAIN',
      'lineno': 96,
      'module_name': 'tests.test_eris',
      'stack': <class 'list'> [
        '
              err1 = Err("Some FOO error has occurred.")
  
        ',
      ],
    },
  ]
---
# name: test_is_json__NO_CAUSE
  <class 'list'> [
    <class 'dict'> {
      'caused_by': <class 'list'> [
      ],
      'exc_info': <class 'dict'> {
        'exc_msg': 'Something went wrong...',
        'exc_type': "<class 'eris._errors.ErisError'>",
        'exc_value': '
          ErisError::tests.test_eris::test_is_json__NO_CAUSE::33{
            Something went wrong...
          }
        ',
      },
      'file_name': '/eris/tests/test_eris.py',
      'func_name': 'test_is_json__NO_CAUSE',
      'lineno': 33,
      'module_name': 'tests.test_eris',
      'stack': <class 'list'> [
        '
              err = Err(ERROR_MSG)
  
        ',
      ],
    },
  ]
---
# name: test_is_json__ONE_CAUSE
  <class 'list'> [
    <class 'dict'> {
      'caused_by': <class 'list'> [
        <class 'tuple'> (
          "<class 'ZeroDivisionError'>",
          "ZeroDivisionError('division by zero')",
          'division by zero',
        ),
      ],
      'exc_info': <class 'dict'> {
        'exc_msg': 'Something went wrong...',
        'exc_type': "<class 'eris._errors.ErisError'>",
        'exc_value': '
          ErisError::tests.test_eris::test_is_json__ONE_CAUSE::47{
            Something went wrong...
          }
        ',
      },
      'file_name': '/eris/tests/test_eris.py',
      'func_name': 'test_is_json__ONE_CAUSE',
      'lineno': 47,
      'module_name': 'tests.test_eris',
      'stack': <class 'list'> [
        '
                  err = Err(ERROR_MSG).chain(zero_div_error)
  
        ',
        '
            File "/eris/tests/test_eris.py", line 44, in test_is_json__ONE_CAUSE
              x = 1 / 0
  
        ',
      ],
    },
  ]
---
# name: test_is_json__TWO_CAUSE
  <class 'list'> [
    <class 'dict'> {
      'caused_by': <class 'list'> [
        <class 'tuple'> (
          "<class 'RuntimeError'>",
          "RuntimeError('Why would we divide by zero?')",
          'Why would we divide by zero?',
        ),
        <class 'tuple'> (
          "<class 'ZeroDivisionError'>",
          "ZeroDivisionError('division by zero')",
          'division by zero',
        ),
      ],
      'exc_info': <class 'dict'> {
        'exc_msg': 'Something went wrong...',
        'exc_type': "<class 'eris._errors.ErisError'>",
        'exc_value': '
          ErisError::tests.test_eris::test_is_json__TWO_CAUSE::65{
            Something went wrong...
          }
        ',
      },
      'file_name': '/eris/tests/test_eris.py',
      'func_name': 'test_is_json__TWO_CAUSE',
      'lineno': 65,
      'module_name': 'tests.test_eris',
      'stack': <class 'list'> [
        '
                      err = Err(ERROR_MSG).chain(rt_error)
  
        ',
        '
            File "/eris/tests/test_eris.py", line 61, in test_is_json__TWO_CAUSE
              raise RuntimeError(
  
        ',
        '
            File "/eris/tests/test_eris.py", line 57, in test_is_json__TWO_CAUSE
              x = 1 / 0
  
        ',
      ],
    },
  ]
---
# name: test_is_json__TWO_CAUSE_AND_RAISE_SELF
  <class 'list'> [
    <class 'dict'> {
      'caused_by': <class 'list'> [
        <class 'tuple'> (
          "<class 'RuntimeError'>",
          "RuntimeError('Why would we divide by zero?')",
          'Why would we divide by zero?',
        ),
        <class 'tuple'> (
          "<class 'ZeroDivisionError'>",
          "ZeroDivisionError('division by zero')",
          'division by zero',
        ),
      ],
      'exc_info': <class 'dict'> {
        'exc_msg': 'Something went wrong...',
        'exc_type': "<class 'eris._errors.ErisError'>",
        'exc_value': '
          ErisError::tests.test_eris::test_is_json__TWO_CAUSE_AND_RAISE_SELF::84{
            Something went wrong...
          }
        ',
      },
      'file_name': '/eris/tests/test_eris.py',
      'func_name': 'test_is_json__TWO_CAUSE_AND_RAISE_SELF',
      'lineno': 84,
      'module_name': 'tests.test_eris',
      'stack': <class 'list'> [
        '
                      error = ErisError(ERROR_MSG)
  
        ',
        '
            File "/eris/tests/test_eris.py", line 80, in test_is_json__TWO_CAUSE_AND_RAISE_SELF
              raise RuntimeError(
  
        ',
        '
            File "/eris/tests/test_eris.py", line 76, in test_is_json__TWO_CAUSE_AND_RAISE_SELF
              x = 1 / 0
  
        ',
      ],
    },
  ]
---
